File: 04-Portfolio/Python/Ejercicio/portfolio/PortfolioTest.py
#
# Developed by 10Pines SRL
# License: 
# This work is licensed under the 
# Creative Commons Attribution-NonCommercial-ShareAlike 3.0 Unported License. 
# To view a copy of this license, visit http://creativecommons.org/licenses/by-nc-sa/3.0/ 
# or send a letter to Creative Commons, 444 Castro Street, Suite 900, Mountain View, 
# California, 94041, USA.
#  
import unittest
from copy import copy


class AccountTransaction:
    def value(self):
        pass

    @classmethod
    def registerForOn(cls, value, account):
        return account.register(cls(value))


class Deposit(AccountTransaction):
    def __init__(self, value):
        self._value = value

    def value(self):
        return self._value


class Withdraw(AccountTransaction):
    def __init__(self, value):
        self._value = value

    def value(self):
        return self._value


class SummarizingAccount:
    def balance(self):
        pass

    def hasRegistered(self, transaction):
        pass

    def manages(self, account):
        pass

    def transactions(self):
        pass


class ReceptiveAccount(SummarizingAccount):
    def __init__(self):
        self._transactions = []

    def balance(self):
        return reduce(lambda balance, transaction: balance + transaction.value(), self._transactions, 0)

    def register(self, aTransaction):
        self._transactions.append(aTransaction)
        return aTransaction

    def hasRegistered(self, transaction):
        return transaction in self._transactions

    def manages(self, account):
        return self == account

    def transactions(self):
        return copy(self._transactions)


class Portfolio(SummarizingAccount):
    def balance(self):
        raise NotImplementedError()

    def hasRegistered(self, transaction):
        raise NotImplementedError()

    def manages(self, anAccount):
        raise NotImplementedError()

    def transactions(self):
        raise NotImplementedError()

    def addAccount(self, account):
        raise NotImplementedError()

    @classmethod
    def createWith(cls, anAccount, anotherAccount):
        raise NotImplementedError()

    ACCOUNT_ALREADY_MANAGED = "La cuenta ya esta manejada por otro portfolio"


class PortfolioTests(unittest.TestCase):
    def test01ReceptiveAccountHaveZeroAsBalanceWhenCreated(self):
        account = ReceptiveAccount()
        self.assertEquals(0, account.balance())

    def test02DepositIncreasesBalanceOnTransactionValue(self):
        account = ReceptiveAccount()
        Deposit.registerForOn(100, account)

        self.assertEquals(100, account.balance())

    def test03WithdrawDecreasesBalanceOnTransactionValue(self):
        account = ReceptiveAccount()
        Deposit.registerForOn(100, account)
        Withdraw.registerForOn(-50, account)

        self.assertEquals(50, account.balance())

    def test04PortfolioBalanceIsSumOfManagedAccountsBalance(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        complexPortfolio = Portfolio()
        complexPortfolio.addAccount(account1)
        complexPortfolio.addAccount(account2)

        Deposit.registerForOn(100, account1)
        Deposit.registerForOn(200, account2)

        self.assertEquals(300, complexPortfolio.balance())

    def test05PortfolioCanManagePortfolios(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        account3 = ReceptiveAccount()
        complexPortfolio = Portfolio.createWith(account1, account2)
        composedPortfolio = Portfolio.createWith(complexPortfolio, account3)

        Deposit.registerForOn(100, account1)
        Deposit.registerForOn(200, account2)
        Deposit.registerForOn(300, account3)
        self.assertEquals(600, composedPortfolio.balance())

    def test06ReceptiveAccountsKnowsRegisteredTransactions(self):
        account = ReceptiveAccount()
        deposit = Deposit.registerForOn(100, account)
        withdraw = Withdraw.registerForOn(-50, account)

        self.assertTrue(account.hasRegistered(deposit))
        self.assertTrue(account.hasRegistered(withdraw))

    def test07ReceptiveAccountsDoNotKnowNotRegisteredTransactions(self):
        account = ReceptiveAccount()
        deposit = Deposit(100)
        withdraw = Withdraw(-50)

        self.assertFalse(account.hasRegistered(deposit))
        self.assertFalse(account.hasRegistered(withdraw))

    def test08PortofoliosKnowsTransactionsRegisteredByItsManagedAccounts(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        account3 = ReceptiveAccount()
        complexPortfolio = Portfolio.createWith(account1, account2)
        composedPortfolio = Portfolio.createWith(complexPortfolio, account3)

        deposit1 = Deposit.registerForOn(100, account1)
        deposit2 = Deposit.registerForOn(200, account2)
        deposit3 = Deposit.registerForOn(300, account3)

        self.assertTrue(composedPortfolio.hasRegistered(deposit1))
        self.assertTrue(composedPortfolio.hasRegistered(deposit2))
        self.assertTrue(composedPortfolio.hasRegistered(deposit3))

    def test09PortofoliosDoNotKnowTransactionsNotRegisteredByItsManagedAccounts(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        account3 = ReceptiveAccount()
        complexPortfolio = Portfolio.createWith(account1, account2)
        composedPortfolio = Portfolio.createWith(complexPortfolio, account3)

        deposit1 = Deposit(100)
        deposit2 = Deposit(200)
        deposit3 = Deposit(300)

        self.assertFalse(composedPortfolio.hasRegistered(deposit1))
        self.assertFalse(composedPortfolio.hasRegistered(deposit2))
        self.assertFalse(composedPortfolio.hasRegistered(deposit3))

    def test10ReceptiveAccountManageItSelf(self):
        account1 = ReceptiveAccount()

        self.assertTrue(account1.manages(account1))

    def test11ReceptiveAccountDoNotManageOtherAccount(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()

        self.assertFalse(account1.manages(account2))

    def test12PortfolioManagesComposedAccounts(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        account3 = ReceptiveAccount()
        complexPortfolio = Portfolio.createWith(account1, account2)

        self.assertTrue(complexPortfolio.manages(account1))
        self.assertTrue(complexPortfolio.manages(account2))
        self.assertFalse(complexPortfolio.manages(account3))

    def test13PortfolioManagesComposedAccountsAndPortfolios(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        account3 = ReceptiveAccount()
        complexPortfolio = Portfolio.createWith(account1, account2)
        composedPortfolio = Portfolio.createWith(complexPortfolio, account3)

        self.assertTrue(composedPortfolio.manages(account1))
        self.assertTrue(composedPortfolio.manages(account2))
        self.assertTrue(composedPortfolio.manages(account3))
        self.assertTrue(composedPortfolio.manages(complexPortfolio))

    def test14AccountsKnowsItsTransactions(self):
        account1 = ReceptiveAccount()

        deposit1 = Deposit.registerForOn(100, account1)

        self.assertEquals(1, len(account1.transactions()))
        self.assertTrue(deposit1 in account1.transactions())

    def test15PortfolioKnowsItsAccountsTransactions(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        account3 = ReceptiveAccount()
        complexPortfolio = Portfolio.createWith(account1, account2)
        composedPortfolio = Portfolio.createWith(complexPortfolio, account3)

        deposit1 = Deposit.registerForOn(100, account1)
        deposit2 = Deposit.registerForOn(200, account2)
        deposit3 = Deposit.registerForOn(300, account3)

        self.assertEquals(3, len(composedPortfolio.transactions()))
        self.assertTrue(deposit1 in composedPortfolio.transactions())
        self.assertTrue(deposit2 in composedPortfolio.transactions())
        self.assertTrue(deposit3 in composedPortfolio.transactions())

    def test16CanNotCreatePortfoliosWithRepeatedAccount(self):
        account1 = ReceptiveAccount()
        try:
            Portfolio.createWith(account1, account1)
            self.fail()
        except Exception as invalidPortfolio:
            self.assertEquals(Portfolio.ACCOUNT_ALREADY_MANAGED, invalidPortfolio.message)

    def test17CanNotCreatePortfoliosWithAccountsManagedByOtherManagedPortfolio(self):
        account1 = ReceptiveAccount()
        account2 = ReceptiveAccount()
        complexPortfolio = Portfolio.createWith(account1, account2)
        try:
            Portfolio.createWith(complexPortfolio, account1)
            self.fail()
        except Exception as invalidPortfolio:
            self.assertEquals(Portfolio.ACCOUNT_ALREADY_MANAGED, invalidPortfolio.message)

File: 04-Portfolio/C#/Ejercicio/Ejercicio/Patterns-Portfolio-Exercise-WithAccountImplementation/PortfolioTest.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Patterns_Portfolio_Exercise_WithAccountImplementation
{
    [TestClass]
    public class PortfolioTest
    {
        [TestMethod]
        public void test01ReceptiveAccountHaveZeroAsBalanceWhenCreated()
        {
            ReceptiveAccount account = new ReceptiveAccount();

            Assert.AreEqual(0.0, account.balance());
        }

        [TestMethod]
        public void test02DepositIncreasesBalanceOnTransactionValue()
        {
            ReceptiveAccount account = new ReceptiveAccount();
            Deposit.registerForOn(100, account);

            Assert.AreEqual(100.0, account.balance());

        }

        [TestMethod]
        public void test03WithdrawDecreasesBalanceOnTransactionValue()
        {
            ReceptiveAccount account = new ReceptiveAccount();
            Deposit.registerForOn(100, account);
            Withdraw.registerForOn(-50, account);

            Assert.AreEqual(50.0, account.balance());
        }

        [TestMethod]
        public void test04PortfolioBalanceIsSumOfManagedAccountsBalance()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);

            Deposit.registerForOn(100, account1);
            Deposit.registerForOn(200, account2);

            Assert.AreEqual(300.0, complexPortfolio.balance());
        }

        [TestMethod]
        public void test05PortfolioCanManagePortfolios()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            Portfolio composedPortfolio = Portfolio.createWith(complexPortfolio, account3);

            Deposit.registerForOn(100, account1);
            Deposit.registerForOn(200, account2);
            Deposit.registerForOn(300, account3);
            Assert.AreEqual(600.0, composedPortfolio.balance());
        }

        [TestMethod]
        public void test06ReceptiveAccountsKnowsRegisteredTransactions()
        {
            ReceptiveAccount account = new ReceptiveAccount();
            Deposit deposit = Deposit.registerForOn(100, account);
            Withdraw withdraw = Withdraw.registerForOn(-50, account);

            Assert.IsTrue(account.registers(deposit));
            Assert.IsTrue(account.registers(withdraw));
        }

        [TestMethod]
        public void test07ReceptiveAccountsDoNotKnowNotRegisteredTransactions()
        {
            ReceptiveAccount account = new ReceptiveAccount();
            Deposit deposit = new Deposit(100);
            Withdraw withdraw = new Withdraw(-50);

            Assert.IsFalse(account.registers(deposit));
            Assert.IsFalse(account.registers(withdraw));
        }

        [TestMethod]
        public void test08PortofoliosKnowsTransactionsRegisteredByItsManagedAccounts()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            Portfolio composedPortfolio = Portfolio.createWith(complexPortfolio, account3);

            Deposit deposit1 = Deposit.registerForOn(100, account1);
            Deposit deposit2 = Deposit.registerForOn(200, account2);
            Deposit deposit3 = Deposit.registerForOn(300, account3);

            Assert.IsTrue(composedPortfolio.registers(deposit1));
            Assert.IsTrue(composedPortfolio.registers(deposit2));
            Assert.IsTrue(composedPortfolio.registers(deposit3));
        }

        [TestMethod]
        public void test09PortofoliosDoNotKnowTransactionsNotRegisteredByItsManagedAccounts()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            Portfolio composedPortfolio = Portfolio.createWith(complexPortfolio, account3);

            Deposit deposit1 = new Deposit(100);
            Deposit deposit2 = new Deposit(200);
            Deposit deposit3 = new Deposit(300);

            Assert.IsFalse(composedPortfolio.registers(deposit1));
            Assert.IsFalse(composedPortfolio.registers(deposit2));
            Assert.IsFalse(composedPortfolio.registers(deposit3));
        }

        [TestMethod]
        public void test10ReceptiveAccountManageItSelf()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();

            Assert.IsTrue(account1.manages(account1));
        }

        [TestMethod]
        public void test11ReceptiveAccountDoNotManageOtherAccount()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();

            Assert.IsFalse(account1.manages(account2));
        }

        [TestMethod]
        public void test12PortfolioManagesComposedAccounts()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);

            Assert.IsTrue(complexPortfolio.manages(account1));
            Assert.IsTrue(complexPortfolio.manages(account2));
            Assert.IsFalse(complexPortfolio.manages(account3));
        }

        [TestMethod]
        public void test13PortfolioManagesComposedAccountsAndPortfolios()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            Portfolio composedPortfolio = Portfolio.createWith(complexPortfolio, account3);

            Assert.IsTrue(composedPortfolio.manages(account1));
            Assert.IsTrue(composedPortfolio.manages(account2));
            Assert.IsTrue(composedPortfolio.manages(account3));
            Assert.IsTrue(composedPortfolio.manages(complexPortfolio));
        }

        [TestMethod]
        public void test14AccountsKnowsItsTransactions()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();

            Deposit deposit1 = Deposit.registerForOn(100, account1);

            Assert.AreEqual(1, account1.transactions().Count);
            Assert.IsTrue(account1.transactions().Contains(deposit1));
        }

        [TestMethod]
        public void test15PortfolioKnowsItsAccountsTransactions()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            Portfolio composedPortfolio = Portfolio.createWith(complexPortfolio, account3);

            Deposit deposit1 = Deposit.registerForOn(100, account1);
            Deposit deposit2 = Deposit.registerForOn(200, account2);
            Deposit deposit3 = Deposit.registerForOn(300, account3);

            Assert.AreEqual(3, composedPortfolio.transactions().Count);
            Assert.IsTrue(composedPortfolio.transactions().Contains(deposit1));
            Assert.IsTrue(composedPortfolio.transactions().Contains(deposit2));
            Assert.IsTrue(composedPortfolio.transactions().Contains(deposit3));
        }

        [TestMethod]
        public void test16PortofolioKnowsItsAccountsTransactions()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            Portfolio composedPortfolio = Portfolio.createWith(complexPortfolio, account3);

            Deposit deposit1 = Deposit.registerForOn(100, account1);

            Assert.AreEqual(1, composedPortfolio.transactionsOf(account1).Count);
            Assert.IsTrue(composedPortfolio.transactionsOf(account1).Contains(deposit1));
        }

        [TestMethod]
        public void test17PortofolioKnowsItsPortfoliosTransactions()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            Portfolio composedPortfolio = Portfolio.createWith(complexPortfolio, account3);

            Deposit deposit1 = Deposit.registerForOn(100, account1);
            Deposit deposit2 = Deposit.registerForOn(100, account2);
            Deposit.registerForOn(100, account3);

            Assert.AreEqual(2, composedPortfolio.transactionsOf(complexPortfolio).Count);
            Assert.IsTrue(composedPortfolio.transactionsOf(complexPortfolio).Contains(deposit1));
            Assert.IsTrue(composedPortfolio.transactionsOf(complexPortfolio).Contains(deposit2));
        }

        [TestMethod]
        public void test18PortofolioCanNotAnswerTransactionsOfNotManagedAccounts()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            ReceptiveAccount account3 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);

            try
            {
                complexPortfolio.transactionsOf(account3);
                Assert.Fail();
            }
            catch (Exception accountNotManaged)
            {
                Assert.AreEqual(Portfolio.ACCOUNT_NOT_MANAGED, accountNotManaged.Message);
            }
        }

        [TestMethod]
        public void test19CanNotCreatePortfoliosWithRepeatedAccount()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            try
            {
                Portfolio.createWith(account1, account1);
                Assert.Fail();
            }
            catch (Exception invalidPortfolio)
            {
                Assert.AreEqual(Portfolio.ACCOUNT_ALREADY_MANAGED, invalidPortfolio.Message);
            }

        }

        [TestMethod]
        public void test20CanNotCreatePortfoliosWithAccountsManagedByOtherManagedPortfolio()
        {
            ReceptiveAccount account1 = new ReceptiveAccount();
            ReceptiveAccount account2 = new ReceptiveAccount();
            Portfolio complexPortfolio = Portfolio.createWith(account1, account2);
            try
            {
                Portfolio.createWith(complexPortfolio, account1);
                Assert.Fail();
            }
            catch (Exception invalidPortfolio)
            {
                Assert.AreEqual(Portfolio.ACCOUNT_ALREADY_MANAGED, invalidPortfolio.Message);
            }
        }

    }
}
